MalomAPI_cpp/perfect_api.cpp

```diff
@@ -0,0 +1,219 @@
+
+#include "perfect_api.h"
+#include "perfect_player.h"
+#include "perfect_game_state.h"
+
+#include <filesystem>
+
+PerfectPlayer *MalomSolutionAccess::pp = nullptr;
+std::exception *MalomSolutionAccess::lastError = nullptr;
+
+int MalomSolutionAccess::getBestMove(int whiteBitboard, int blackBitboard,
+                                     int whiteStonesToPlace,
+                                     int blackStonesToPlace, int playerToMove,
+                                     bool onlyStoneTaking)
+{
+    initializeIfNeeded();
+
+    GameState s;
+
+    const int W = 0;
+    const int B = 1;
+
+    if ((whiteBitboard & blackBitboard) != 0) {
+        throw std::invalid_argument("whiteBitboard and blackBitboard shouldn't "
+                                    "have any overlap");
+    }
+
+    for (int i = 0; i < 24; i++) {
+        if ((whiteBitboard & (1 << i)) != 0) {
+            s.T[i] = W;
+            s.stoneCount[W] += 1;
+        }
+        if ((blackBitboard & (1 << i)) != 0) {
+            s.T[i] = B;
+            s.stoneCount[B] += 1;
+        }
+    }
+
+    s.phase = ((whiteStonesToPlace == 0 && blackStonesToPlace == 0) ? 2 : 1);
+    mustBeBetween("whiteStonesToPlace", whiteStonesToPlace, 0, Rules::maxKSZ);
+    mustBeBetween("blackStonesToPlace", blackStonesToPlace, 0, Rules::maxKSZ);
+    s.setStoneCount[W] = Rules::maxKSZ - whiteStonesToPlace;
+    s.setStoneCount[B] = Rules::maxKSZ - blackStonesToPlace;
+    s.kle = onlyStoneTaking;
+    mustBeBetween("playerToMove", playerToMove, 0, 1);
+    s.sideToMove = playerToMove;
+    s.moveCount = 10;
+
+    if (s.futureStoneCount(W) > Rules::maxKSZ) {
+        throw std::invalid_argument("Number of stones in whiteBitboard + "
+                                    "whiteStonesToPlace > " +
+                                    std::to_string(Rules::maxKSZ));
+    }
+    if (s.futureStoneCount(B) > Rules::maxKSZ) {
+        throw std::invalid_argument("Number of stones in blackBitboard + "
+                                    "blackStonesToPlace > " +
+                                    std::to_string(Rules::maxKSZ));
+    }
+
+    std::string errorMsg = s.setOverAndCheckValidSetup();
+    if (errorMsg != "") {
+        throw std::invalid_argument(errorMsg);
+    }
+    if (s.over) {
+        throw std::invalid_argument("Game is already over.");
+    }
+
+    s.lastIrrev = 0;
+
+    int ret = 0;
+
+    try {
+        ret = pp->chooseRandom(pp->goodMoves(s)).toBitBoard();
+    } catch (std::out_of_range &) {
+        throw std::runtime_error("We don't have a database entry for this "
+                                 "position. This can happen either if the "
+                                 "database is corrupted (missing files), or "
+                                 "sometimes when the position is not reachable "
+                                 "from the starting position.");
+    }
+
+    deinitializeIfNeeded();
```

Comment:
Given the performance-sensitive nature of our applications, we'd recommend against calling `deinitializeIfNeeded()` in every call of `getBestMove`, because in this case the `initializeIfNeeded()` call at the beginning of the function would have to actually perform the initialization every time, which would have a non-trivial performance cost.